java/src/hu/holyoil/view/panels/EnvironmentPanel.java

```diff
@@ -60,43 +126,104 @@ else if (res.IsSameType(new Water()))
+            Point point = asteroidPointMap.get(asteroid);
+            g.drawImage(asteroidImg, point.x, point.y, 44, 44, this);
+
+            AbstractBaseResource res = asteroid.GetResource();
+            if (res != null)
+                g.drawImage(DefineImageFrom(res), point.x + 2, point.y + 2, 40, 40, this);
+
+            // draw spaceships
+            List<AbstractSpaceship> spaceships = asteroid.GetSpaceships();
+
+            double phi = -0.5 * Math.PI;
+            double deltaPhi = 2 * Math.PI / (double)spaceships.size();
+            int x = (int) (Math.cos(phi) * 32) + point.x + 12;
+            int y = (int) (Math.sin(phi) * 32) + point.y + 12;
+
+            if (asteroid.GetTeleporter() != null) {
+                g.drawImage(teleportImg, x, y, 20, 20, this);
+                deltaPhi = 2 * Math.PI / (double)(spaceships.size() + 1);
+                phi += deltaPhi;
+            }
+
+            for (AbstractSpaceship spaceship : spaceships) {
+                x = (int) (Math.cos(phi) * 32) + point.x + 12;
+                y = (int) (Math.sin(phi) * 32) + point.y + 12;
+                g.drawImage(DefineImageFrom(spaceship), x, y, 20, 20, this);
+                phi += deltaPhi;
+            }
         }
 
-        g.drawImage(asteroidImg, 330, 260, 42, 42, this);
+        // draw central asteroid
+        g.drawImage(asteroidImg, 330, 280, 64, 64, this);
         AbstractBaseResource res = player.GetOnAsteroid().GetResource();
-        if (res == null)
-            return;
-        g.drawImage(DefineImageFrom(res), 331, 261, 40, 40, this);
+        if (res != null)
+            g.drawImage(DefineImageFrom(res), 334, 284, 56, 56, this);
+
+        // draw teleport
+        if (player.GetOnAsteroid().GetTeleporter() != null)
+            g.drawImage(teleportImg, 230, 210, 60, 60, this);
+
+        // draw spaceships
+        List<AbstractSpaceship> spaceships = player.GetOnAsteroid().GetSpaceships();
+        double phi = -0.5 * Math.PI;
+        double deltaPhi = 2 * Math.PI / (double)spaceships.size();
+        for (AbstractSpaceship spaceship : spaceships) {
+            int x = (int) (Math.cos(phi) * 50) + 330 + 12;
+            int y = (int) (Math.sin(phi) * 50) + 280 + 12;
+            g.drawImage(DefineImageFrom(spaceship), x, y, 40, 40, this);
+            phi += deltaPhi;
+        }
     }
 
     @Override
     public void UpdateComponent() {
         player = TurnController.GetInstance().GetSteppingSettler();
         sunstormCountLabel.setText(SunController.GetInstance().GetTurnsUntilStorm() + " turn(s)");
+
+        // define asteroids place
+        List<Asteroid> asteroids = player.GetOnAsteroid().GetNeighbours();
+        double phi = 0;
+        double deltaPhi = 2 * Math.PI / asteroids.size();
+        for (Asteroid asteroid : asteroids) {
+            int x = (int) (Math.cos(phi) * 280) + 330;
+            int y = (int) (Math.sin(phi) * 220) + 280;
+            asteroidPointMap.put(asteroid, new Point(x, y));
```

Comment:
You use foreach on `asteroidPointMap`  when initializing the listeners. You should clear the map in the beginning of `UpdateComponent`, because moving with the player will pollute it with unwanted hitboxes.

Reply:
Done